src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,257 @@
+        {
+            var request = new InitiateMultipartUploadRequest {BucketName = bucketName, Key = artifactSettings.Name, ContentType = artifactSettings.ContentType};
+            foreach (var metaData in artifactSettings.Metadata)
+                request.Metadata[metaData.Key] = metaData.Value;
+            var response = await client.InitiateMultipartUploadAsync(request, cancellationToken);
+            return response.UploadId;
+        }
+
+        private static async Task CompleteMultiPartUploadAsync(IAmazonS3 client, string bucketName, string uploadId, EgressArtifactSettings artifactSettings, List<PartETag> parts, CancellationToken cancellationToken)
+        {
+            var request = new CompleteMultipartUploadRequest
+            {
+                BucketName = bucketName,
+                Key = artifactSettings.Name,
+                UploadId = uploadId,
+                PartETags = parts
+            };
+            await client.CompleteMultipartUploadAsync(request, cancellationToken);
+        }
+
+        private static T WrapException<T>(Func<T> func)
+        {
+            try
+            {
+                return func();
+            }
+            catch (DirectoryNotFoundException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (PathTooLongException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (IOException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (NotSupportedException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (SecurityException ex)
+            {
+                throw CreateException(ex);
+            }
+            catch (UnauthorizedAccessException ex)
+            {
+                throw CreateException(ex);
+            }
+        }
+
+        private static EgressException CreateException(string message)
+        {
+            return new EgressException(WrapMessage(message));
+        }
+
+        private static EgressException CreateException(Exception innerException)
+        {
+            return new EgressException(WrapMessage(innerException.Message), innerException);
+        }
+
+        private static string WrapMessage(string innerMessage)
+        {
+            return !string.IsNullOrEmpty(innerMessage) 
+                ? string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressFileFailedDetailed, innerMessage) 
+                : Strings.ErrorMessage_EgressFileFailedGeneric;
+        }
+
+        private static async Task VerifyBucketExistsAsync(IAmazonS3 client, string bucketName, CancellationToken token)
+        {
+            var response = await client.ListBucketsAsync(token);
+            if (response.Buckets.Any(b => string.Equals(b.BucketName, bucketName, StringComparison.OrdinalIgnoreCase)))
```

Comment:
you can use DoesS3BucketExistV2Async api => https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/S3/MS3UtilDoesS3BucketExistV2AsyncIS3String.html